Copied tile flip/rotate from Tiled

Ohmnivore opened the pull request:
For feature request #128

Added persistent tile meta-data (TileData class) in the process which could be helpful for future tile features.

Use H/V/R/E keys to respectively flip brush horizontally, flip brush vertically, rotate brush clockwise, and rotate brush counter-clockwise.

For export it uses the highest 3 bits (signed 32 bit int) of the tile index to store flip/rotate info. For COORD export mode it uses the highest 3 bits of the x component, y is untouched. This isn't a breaking change for the format until flip/rotate is actively used in a level.

Perhaps an explicit project or layer setting to enable flip/rotate could be useful.

Tested all tile tools to the best of my ability.

Here are the references I used from Tiled:
1. https://doc.mapeditor.org/en/stable/reference/tmx-map-format/#tile-flipping
2. https://discourse.mapeditor.org/t/can-i-rotate-tiles/703/6
3. https://github.com/bjorn/tiled/issues/2119
4. https://github.com/bjorn/tiled/blob/1f5185e01aa3cbc60571ca9ffa4377d71f400ef9/src/libtiled/tilelayer.cpp#L394

Ohmnivore commented on src/rendering/GLRenderer.hx:
```diff
@@ -277,27 +278,60 @@ class GLRenderer
+		var topLeft = new Vector(-1, -1);
+		var topRight = new Vector(1, -1);
+		var botLeft = new Vector(-1, 1);
+		var botRight = new Vector(1, 1);
```
Should probably define these and other temp Vectors as class members instead of locals

01010111 commented:
I'm a little worried that using the tile IDs to store data might not be intuitive for people. Since we rely on the community to make importers for different engines it might be a little weird for them to do bitmasking and stuff for implementation. Could translation data be stored somewhere else?

01010111 commented:
I'm getting an error when loading an older JSON file with this branch:
```
io/LevelManager.hx:75: TypeError: Cannot read property '0' of undefined
    at modules_tiles_TileLayer.load (http://localhost:8080/ogmo.js:40265:41)
    at level_data_Level.load (http://localhost:8080/ogmo.js:31076:12)
    at new level_data_Level (http://localhost:8080/ogmo.js:31047:8)
    at Function../ogmo.hxml.io_Imports.levelJSON (http://localhost:8080/ogmo.js:30508:9)
    at Function../ogmo.hxml.io_Imports.level (http://localhost:8080/ogmo.js:30485:20)
    at http://localhost:8080/ogmo.js:30606:24
    at io_LevelManager.enforceLimit (http://localhost:8080/ogmo.js:30740:4)
    at io_LevelManager.open (http://localhost:8080/ogmo.js:30604:8)
    at level_editor_ui_LevelsPanel.selectLevel (http://localhost:8080/ogmo.js:33423:29)
    at HTMLDivElement.<anonymous> (http://localhost:8080/ogmo.js:45101:12)
```
the level:
```json
{
  "ogmoVersion": "3.3.0",
  "width": 320,
  "height": 256,
  "offsetX": 0,
  "offsetY": 0,
  "layers": [
    {
      "name": "tiles",
      "_eid": "07074416",
      "offsetX": 0,
      "offsetY": 0,
      "gridCellWidth": 16,
      "gridCellHeight": 16,
      "gridCellsX": 20,
      "gridCellsY": 16,
      "tileset": "tiles",
      "data2D": [
        [132, 177, 165, 177, 178, 177, 164, 177, 133, 132, 177, 177, 165, 177, 178, 179, 208, 145, 145, 145],
        [147, 0, 16, 0, 17, 0, 16, 0, 144, 147, 96, 81, 16, 2, 18, 3, 208, 209, 145, 145],
        [147, 65, 1, 18, 0, 16, 1, 16, 144, 147, 87, 16, 0, 16, 0, 19, 224, 196, 209, 145],
        [163, 96, 69, 67, 16, 0, 66, 67, 144, 147, 83, 0, 136, 137, 139, 3, 51, 224, 225, 225],
        [147, 96, 96, 96, 68, 69, 96, 83, 144, 147, 2, 16, 152, 154, 155, 16, 21, 22, 21, 128],
        [147, 96, 84, 85, 84, 85, 83, 0, 144, 147, 16, 0, 168, 153, 171, 0, 16, 1, 16, 144],
        [147, 83, 0, 16, 0, 18, 0, 16, 176, 179, 0, 18, 184, 185, 187, 16, 2, 16, 0, 160],
        [163, 0, 16, 0, 16, 0, 17, 0, 18, 0, 16, 0, 16, 2, 16, 0, 16, 64, 65, 144],
        [147, 16, 2, 18, 0, 16, 0, 16, 0, 16, 0, 18, 0, 16, 0, 17, 64, 96, 87, 144],
        [147, 0, 16, 0, 16, 0, 136, 139, 128, 130, 130, 130, 131, 0, 16, 128, 129, 129, 129, 145],
        [147, 65, 0, 16, 0, 16, 184, 187, 144, 132, 177, 177, 179, 16, 0, 176, 177, 177, 177, 133],
        [163, 96, 65, 0, 17, 0, 16, 0, 144, 147, 16, 0, 16, 0, 18, 0, 66, 96, 96, 144],
        [147, 96, 96, 69, 68, 65, 2, 16, 144, 147, 0, 16, 0, 17, 0, 66, 96, 96, 96, 160],
        [147, 96, 96, 96, 96, 96, 65, 0, 144, 147, 16, 1, 16, 0, 16, 70, 96, 96, 96, 176],
        [147, 96, 96, 96, 96, 96, 96, 69, 144, 147, 0, 16, 0, 16, 64, 96, 96, 192, 193, 193],
        [148, 129, 130, 130, 130, 130, 129, 130, 149, 148, 129, 129, 129, 129, 129, 129, 131, 208, 209, 209]
      ],
      "exportMode": 0,
      "arrayMode": 1
    }
  ]
}
```

Ohmnivore commented:
Fixed the error above.

Also adopted new export/import strategy:

Added `tileFlags`/`tileFlags2D` (depends on array mode). The lower three bits are used to store horizontal, vertical, and diagonal flip respectively.

If a tile layer has no flipped or rotated tiles then the `tileFlags`/`tileFlags2D` arrays are exported as empty to save space.

I experimented with exporting rotation instead of diagonal flip but the bit wrangling and conversion for that case led to more complex code so I opted the other way.

A reference `TileFlipRotationData.fromFlags(flags:Int):TileFlipRotationData` function implementation is provided to convert from diagonal flip to rotation (for parser libraries).

AustinEast commented:
this seems good to me! is your issue fixed @01010111?
